### ISSUE TITLE
[Artery] Memory leak due to compression keeping StreamRefs after stream is complete

### ISSUE THREAD

**Issue by nodefactory-bk** [akka/akka#28959]
Artery actorref compression seems to keep reference to streamref actors in the heavyhitters-map long after the stream itself has terminated.

This causes the streams and their parts not to be garbage collected properly.

This is likely something similar to 
* https://github.com/akka/akka/pull/28528
* https://github.com/akka/akka/issues/28822

Terminating the association in question immediately releases all memory held and the no longer used streams and their producers get GC'd
The problem doesn't appear when using classic remoting.

Akka versions tested with are 2.5.31 and 2.5.29. Can't yet migrate to 2.6.x so haven't tested 2.6.4.

Here's a slightly anonymized example of an compression advertisement.
There were no streams running at the time.

```
Incoming ActorRef compression advertisement from [akka://prod@10.12.9.13:2552#-2659742560248075754], table: [CompressionTable(7526929967489218174,56,HashMap(Actor[akka://prod@10.12.9.13:2552/system/distributedPubSubMediator#-1851829994] -> 5, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$Cjc-SinkRef-19276#-1290794736] -> 23, Actor[akka://prod@10.12.9.13:2552/] -> 29, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$FBc-SinkRef-21613#770586671] -> 25, Actor[akka://prod/system/cluster/core/daemon#1092588098] -> 16, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$y7b-SourceRef-161#956005202] -> 3, Actor[akka://prod/user/tenantshardresolver#553133675] -> 14, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$dy-SinkRef-3851#201691900] -> 2, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4#-1638701235] -> 6, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$+p-SourceRef-15#-147374760] -> 17, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$HUc-SinkRef-24086#1738433618] -> 27, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$XTb-SourceRef-135#2011851815] -> 9, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$Ifc-SourceRef-184#1327161540] -> 28, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$1n-SinkRef-2447#-1082149150] -> 18, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$6M-SinkRef-5927#-1402474863] -> 19, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$X+b-SourceRef-168#-1344279657] -> 26, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$ncc-SourceRef-177#-722950990] -> 12, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$Vac-SinkRef-17975#-1062652880] -> 22, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$~2b-SinkRef-17298#1033459610] -> 15, Actor[akka://prod@10.12.9.13:2552/system/distributedPubSubMediator/tenantshard#615398113] -> 0, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$TJb-SourceRef-119#847832134] -> 8, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$K0b-SourceRef-144#-188053557] -> 24, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$lJ-SourceRef-36#-1666246589] -> 20, Actor[akka://prod@10.12.9.13:2552/system/cluster/core/daemon#-135251251] -> 1, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$HFc-SinkRef-22445#-776927854] -> 11, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$4Ic-SinkRef-22733#-1720755280] -> 4, Actor[akka://prod/system/distributedPubSubMediator#1757190690] -> 13, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$Uz-SourceRef-26#790686326] -> 7, Actor[akka://prod/user/Powell/http/StreamSupervisor-4/$$cNb-SourceRef-125#2068300073] -> 21, Actor[akka://prod@10.12.9.13:2552/user/tenantshard-4/***/product/productdb/StreamSupervisor-607/$$3qc-SinkRef-20332#509009238] -> 10))]
```

btw, is there a way to completely disable the compression? Can't seem to find a setting to disable it.

**Comment by patriknw** [comment]
Strange that the streams can't be garbage collected because of keeping the ActorRef around. That's something we should look into.

Probably undocumented but compression can be disabled with config:
```
akka.remote.artery.advanced.compression {
  actor-refs.max = 0
}
```

**Comment by nodefactory-bk** [comment]
If it helps here are some screenshots from a heap dump

First screenshot attempts to show one of the cached stream refs. The second is the highlighted functionref's gc root path.
![scr1](https://user-images.githubusercontent.com/16016450/79975499-5fc89680-849b-11ea-81f3-173642bbb0a3.png)
![scr2](https://user-images.githubusercontent.com/16016450/79975501-60f9c380-849b-11ea-94ad-e21e194ed418.png)

**Comment by nodefactory-bk** [comment]
When setting actor-refs.max = 1 the "sending" end of the stream is better behaved.
However the receiver still fills up the "ActorRefResolveCacheWithAddress" cache with these.
And that cache is hard coded to 1024 entries.

Here's a few screenshots from MemoryAnalyzer for this:
![decoder](https://user-images.githubusercontent.com/16016450/80960090-75797c80-8e08-11ea-95bd-c34901779435.png)

**Comment by patriknw** [comment]
@nodefactory-bk I'm pretty sure my fix in https://github.com/akka/akka/pull/29026 will help but I couldn't reproduce the scenario with StreamRefs. Maybe we can build a snapshot version that you can test in your application?

**Comment by nodefactory-bk** [comment]
I can build the snapshots locally so that's no issue.
I'll try the change and report back in a few days or so.

**Comment by patriknw** [comment]
Thanks

**Comment by johanandren** [comment]
Sorry, I missed we were waiting for feedback on that and merged the PR into master.

**Comment by patriknw** [comment]
np,  it should be good. if something more we can follow up

**Comment by nodefactory-bk** [comment]
I cherry-picked the fix on top of 2.5.31 and have been running it in my test envs.
As far as I can tell this has completely solved the memory issues.

Thanks!

**Comment by patriknw** [comment]
Thanks for verifying. I hope you can update to Akka 2.6.x
